lib/scan-includes.js

```diff
@@ -0,0 +1,110 @@
+const realFetchUrl = require("./fetch-url");
+const GITHUB_SERVICE = require("./services").GITHUB;
+
+
+/** Scans for recursively included paths from a root specification.
+ *
+ * @param {string} repositoryUrl The base URL of the repository. The function
+ * won't fetch anything outside of this scope. This must end with a `/`.
+ * @param {string} rootSpec A path relative to the `repositoryUrl` pointing to
+ * the root specification in the `specType` format.
+ * @param {"bikeshed"|"respec"} specType
+ *
+ * @returns {Promise<Set<string>>} Paths relative to `repositoryUrl` that are
+ * recursively included by the specification.
+ */
+module.exports = async function scanIncludes(repositoryUrl, rootUrl, specType, fetchUrl = realFetchUrl) {
+    if (!repositoryUrl.endsWith('/')) {
+        throw new Error(`${repositoryUrl} must end with a '/'.`);
+    }
+    let includeRE;
+    switch (specType) {
+        case "bikeshed":
+            includeRE = /^path:([^\n]+)$/gm;
+            break;
+        case "respec":
+            includeRE = /data-include="([^"]+)"|data-include='([^']+)'/g;
+            break;
+        default:
+            throw new Error(`Unexpected specification type: ${JSON.stringify(specType)}`);
+    }
+
+    /** All the files that successfully fetched go here. */
+    const recursiveIncludes = new Set();
+    /** When this gets back to empty, we're done searching includes. */
+    const inProgressFetches = new Set();
+
+    await new Promise((resolve, reject) => {
```

Comment:
At first glance, this promise seems redundant as it doesn't operate on a callback... just `await scanOneFile()` and get it to throw as normal. You can probably also extract `scanOneFile()` outside of this function and get it to return `recursiveIncludes ` ... they could be passed as default arguments... something like: 

```js
scanOneFile(path);
async function scanOneFile(path, inProgressFetches = new Set(), recursiveIncludes = new Set()) {
    // do sync stuff... can throw 
    // do recursion:
    await scanOneFile(path, inProgressFetches, recursiveIncludes);
    // eventually
    return recursiveIncludes;
} ;
```

Reply:
Good point. I was worried about allowing maximal parallelism in the fetches, but a `Promise.all()` seems to allow the same amount, and it's much simpler to read.

Doing the accumulator sets as parameters initialized by default arguments seems less clear, to me, than making it a nested function. Similarly, it seems clearer which value is returned if it's returned unconditionally as the last line of `scanIncludes()`, than if readers have to also check what `scanOneFile()` returns.